File: src/birdie_ffi.erl
-module(birdie_ffi).
-export([get_line/1, is_windows/0]).

get_line(Prompt) ->
    case io:get_line(Prompt) of
        eof -> {error, nil};
        {error, _} -> {error, nil};
        Data when is_binary(Data) -> {ok, Data};
        Data when is_list(Data) -> {ok, unicode:characters_to_binary(Data)}
    end.

is_windows() ->
    case os:type() of
        {win32, _} -> true;
        _ -> false
    end.

File: src/birdie_ffi.mjs
import fs from "node:fs";
import { exit } from "node:process";
import { Result$Ok, Result$Error } from "./gleam.mjs";

export function is_windows() {
  return globalThis?.process?.platform === "win32" || globalThis?.Deno?.build?.os === "windows";
}

export function halt(status_code) {
  exit(status_code);
}

export function get_line(prompt) {
  process.stdout.write(prompt);
  const buffer = Buffer.alloc(4096);
  while (true) {
    try {
      const bytesRead = fs.readSync(process.stdin.fd, buffer, 0, buffer.length, null);
      const input = buffer.toString("utf-8", 0, bytesRead);
      return Result$Ok(input);
    } catch (error) {
      if (error.code === "EAGAIN") continue;
      return Result$Error(undefined);
    }
  }
}
